doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -1,209 +1,213 @@
+- :gdef:`Setoid equality <setoid equality>` defines equality in terms of an equivalence
+  relation.  A :gdef:`setoid` is a set that is equipped with an equivalence relation
+  (see https://en.wikipedia.org/wiki/Setoid).  These are needed to form a :gdef:`quotient set`
+  or :gdef:`quotient`
+  (see https://en.wikipedia.org/wiki/Equivalence_Class).  In Coq, users generally work
+  with setoids rather than constructing quotients, for which there is no specific support.
 
-   where :g:`eq` is the Leibniz equality or a registered setoid equality.
+- :gdef:`Definitional equality <definitional equality>` is equality based on the conversion
+  rules, which Coq can determine automatically.  (See :ref:`Conversion-rules`)
+  When two terms are definitionally equal, Coq knows it can
+  replace one with the other, such as with :tacn:`change` `X with Y`, among many
+  other advantages.  ":term:`Convertible <convertible>`" is another way of saying that
+  two terms are definitionally equal.
 
-   Then :n:`rewrite @term` finds the first subterm matching `term`\ :sub:`1` in the goal,
-   resulting in instances `term`:sub:`1`' and `term`:sub:`2`' and then
-   replaces every occurrence of `term`:subscript:`1`' by `term`:subscript:`2`'.
-   Hence, some of the variables :g:`x`\ :sub:`i` are solved by unification,
-   and some of the types :g:`A`\ :sub:`1`:g:`, ..., A`\ :sub:`n` become new
-   subgoals.
+  .. todo PR: A and B are convertible iff they are definitionally equal??
 
-   .. exn:: The @term provided does not end with an equation.
-      :undocumented:
+.. _rewritingexpressions:
 
-   .. exn:: Tactic generated a subgoal identical to the original goal. This happens if @term does not occur in the goal.
-      :undocumented:
+Rewriting with Leibniz and setoid equality
+------------------------------------------
 
-   .. tacv:: rewrite -> @term
+.. tacn:: rewrite {+, @oriented_rewriter } {? @occurrences } {? by @ltac_expr3 }
 
-      Is equivalent to :n:`rewrite @term`
+   .. insertprodn oriented_rewriter one_term_with_bindings
 
-   .. tacv:: rewrite <- @term
+   .. prodn::
+      oriented_rewriter ::= {? {| -> | <- } } {? @natural } {? {| ? | ! } } @one_term_with_bindings
+      one_term_with_bindings ::= {? > } @one_term {? with @bindings }
 
-      Uses the equality :n:`@term`:sub:`1` :n:`= @term` :sub:`2` from right to left
+   .. todo PR: When I search for "rewrite (" in the .v files, I see tons of examples that I don't
+      understand.  Need to say something about them.
 
-   .. tacv:: rewrite @term in @goal_occurrences
+   Rewrites terms based on equalities.  :n:`@one_term` can be the name of a hypothesis that has
+   the form :n:`EQ @term__1 @term__2` or a lemma that has the form:
 
-      Analogous to :n:`rewrite @term` but rewriting is done following
-      the clause :token:`goal_occurrences`. For instance:
+   :n:`forall (x__1: A__1) ... (x__n: A__n), EQ @term__1 @term__2`
```

Comment:
What about:
```suggestion
   :n:`{* forall {+ @binder }, } EQ @term__1 @term__2`
```

It would be more general, although it would need some rewrite in the following text.

Another option requiring less rewrite is:
```suggestion
   :n:`{* forall {+ (x__i: A__i) }, } EQ @term__1 @term__2`
```

Reply:
Are all the options of `binder` supported here?  I expect the common case is just `(x__i : A__i)`.  I have this now:

![image](https://user-images.githubusercontent.com/1253341/101728100-a1c63d00-3a6a-11eb-9fb9-3e93f49840a9.png)

I could just say that this is the most common form and that you can, in fact, use any type of `binder` in place of `(x__i : A__i)`.

Reply:
> I could just say that this is the most common form and that you can, in fact, use any type of `binder` in place of `(x__i : A__i)`.

Sounds good to me (I think that's the truth).